README.md
# happner-files

Happner file server.

Currently only supports uploads.

## Configuration

In mesh config.

```javascript
  ...
  components: {
    'happner-files': {
      // define a set of url to fs mappings
      // (first match wins)
      path: {
        routes: {
          '/match/url/path': '/store/here',
          '/': '/var/www'
        }
      },
      // define the happner web route to the upload handler
      web: {
        routes: {
          // http://host.name/happner-files/files
          files: 'handler'
        }
      }
    }
  },
  ...
```

An upload POST to */happner-files/files***/match/url/path**/filename<br/>
will be stored at **/store/here**/filename,<br/>
having matched the first *path route*.

An upload POST to */happner-files/files* **/** kept/deeper/path/filename<br/>
will be stored at **/var/www**/kept/deeper/path/filename,<br/>
having matched the second *path route*.


RELEASES.md
## 0.1.0
- support GET

## 0.0.2
- support POST
- add example web page posting to secure server

## 1.0.0
- happner-2

## 1.1.0
- SMC-3379: cross origin headers, support for OPTIONS


test/unit-files.js
var Files = require('../lib/files');
var should = require('should');
var path = require('path');
var fs = require('fs');
var rimraf = require('rimraf');

describe('files', function() {
  beforeEach(function() {
    this.files = new Files();
  });

  beforeEach(function() {
    this.$happn = {
      config: {
      },
      log: {
        $$DEBUG: function() {
          // console.log(arguments);
        },
        error: function() {
        }
      }
    };
    this.req = {
      url: '/the/url'
    };
    this.res = {
      end: function() {}
    };
  });


  context('routes', function() {
    it('has an empty set of routes upon initialization', function() {
      this.files.routes.should.eql([]);
    });
  });

  context('handler()', function() {

    it('is defined', function() {
      this.files.handler.should.be.an.instanceof(Function);
    });

    context('calls _setPathRoutes()', function() {
      it('to load the route regex matchers', function(done) {
        this.files._setPathRoutes = function() {
          done();
          return true;
        };
        this.files.handler(this.$happn, this.req, this.res);
      });
    });

    context('calls _matchPathRoute() with the url', function() {
      it('to get the matching path route', function(done) {
        this.files._setPathRoutes = function() {
          return true;
        };
        this.files._matchPathRoute = function(url) {
          url.should.equal('/the/url');
          done();
        };
        this.files.handler(this.$happn, this.req, this.res);
      });
    });

    it('calls the appropriate METHOD handler', function(done) {
      this.files._setPathRoutes = function() {
        return true;
      };
      this.files._matchPathRoute = function() {
        return '/some/new/path';
      };
      this.req.method = 'POST';
      this.files._handlePOST = function($happn, targetFilename, req, res) {
        targetFilename.should.equal('/some/new/path');
        done();
      };
      this.files.handler(this.$happn, this.req, this.res);
    });

    it('responds 500 on not yet implemented METHOD handlers', function(done) {
      this.files._setPathRoutes = function() {
        return true;
      };
      this.files._matchPathRoute = function() {
        return '/some/new/path';
      };
      this.req.method = 'HEAD';
      var res = this.res;
      this.res.end = function(body) {
        res.statusCode.should.equal(500);
        body.should.equal('happner-files: HEAD method not implemented');
        done();
      };
      this.files.handler(this.$happn, this.req, this.res);
    });

  });

  context('_setPathRoutes()', function() {
    it('returns false if no routes config', function() {
      var $happn = {
        config: {}
      };
      this.files._setPathRoutes($happn).should.equal(false);
    });

    it('returns true if routes already exist on "this"', function() {
      var $happn = {
        config: {}
      };
      this.files.routes = [{match: /xxx/, path: '/path'}];
      this.files._setPathRoutes($happn).should.equal(true);
    });

    it('creates a regex matcher for each route in config', function() {
      var $happn = {
        config: {
          path: {
            routes: {
              '/path': '/tmp/files'
            }
          }
        }
      };
      this.files._setPathRoutes($happn).should.equal(true);
      this.files.routes.should.eql([
        {match: new RegExp('^/path'), path: '/tmp/files'} // mocha comparator needs a dynamic regex here
      ]);
    });
  });

  context('_matchPathRoute()', function() {
    it('returns the targetFilename with the matching path route substituted in', function(done) {
      this.files.routes = [
        {
          match: /^\/path/,
          path: '/tmp/files'
        }
      ];
      var targetFilename = this.files._matchPathRoute('/path/some/name');
      targetFilename.should.equal('/tmp/files/some/name');
      done();
    });

    it('returns the first match', function(done) {
      this.files.routes = [
        {
          match: /^\/harth/,
          path: '/tmp/files'
        },
        {
          match: /^\/path/,
          path: '/tmp/files'
        }
      ];
      var targetFilename = this.files._matchPathRoute('/path/some/name');
      targetFilename.should.equal('/tmp/files/some/name');
      done();
    });

    it('returns false if no match', function(done) {
      this.files.routes = [
        {
          match: /^\/garth/,
          path: '/tmp/files'
        }
      ];
      var targetFilename = this.files._matchPathRoute('/path/some/name');
      targetFilename.should.equal(false);
      done();
    });
  });

  context('_handlePOST', function() {
    it('creates the target directory', function(done) {
      var targetFilename = __dirname + path.sep + 'tmp' + path.sep + 'directory' + path.sep + 'filename';
      this.req.on = function(event, handler) {
        if (event === 'end') handler();
      };
      this.res.end = function() {
        fs.lstatSync(path.dirname(targetFilename)); // should exist, or throws to fail test
        rimraf.sync(path.dirname(targetFilename));
        done();
      };
      this.files._handlePOST(this.$happn, targetFilename, this.req, this.res);
    });
  });

  context('_handleGET', function() {
    it('returns 404 if no such file', function(done) {
      var targetFilename = './some/nonexistant/path';
      delete this.res.statusCode;
      var _this = this;
      this.res.end = function() {
        _this.res.statusCode.should.equal(404);
        done();
      };
      this.files._handleGET(this.$happn, targetFilename, this.req, this.res);
    });

    it('returns 500 if some error other than file not found occurs', function(done) {
      var targetFilename = './the/mocked/fs/lstat/below/causes/other/someerror/on/this/path';
      var original = fs.lstat;
      fs.lstat = function(path, callback) {
        callback(new Error('some other error'));
      };
      delete this.res.statusCode;
      var _this = this;
      this.res.end = function() {
        fs.lstat = original;
        _this.res.statusCode.should.equal(500);
        done();
      };
      this.files._handleGET(this.$happn, targetFilename, this.req, this.res);
    });

    it('returns 404 if file is a directory', function(done) {
      // Will support directory listing later
      var targetFilename = '/directory/mocked/in/fs/lstat/below';
      var original = fs.lstat;
      fs.lstat = function(path, callback) {
        callback(null, {
          isDirectory: function() {
            return true;
          }
        });
      };
      delete this.res.statusCode;
      var _this = this;
      this.res.end = function() {
        fs.lstat = original;
        _this.res.statusCode.should.equal(404);
        done();
      };
      this.files._handleGET(this.$happn, targetFilename, this.req, this.res);
    });
  });

  context('windows', function() {
    it('is supported');
  });

});


test/func-options.js
const Happner = require('happner-2');
const path = require('path');
const should = require('should');
const http = require('http');

describe(path.basename(__filename), function() {

  before(function(done) {
    const _this = this;
    Happner.create({
      name: 'SERVER',
      host: '127.0.0.1',
      port: 8080,
      util: {
        logLevel: 'debug',
        logComponents: ['happner-files']
      },
      modules: {
        'happner-files': {
          path: path.dirname(__dirname)
        }
      },
      components: {
        'happner-files': {
          path: {
            routes: {
              '/': __dirname + path.sep + 'tmp'
            }
          },
          web: {
            routes: {
              files: 'handler'
            }
          }
        }
      }
    }).then(function(mesh) {
      _this.server = mesh;
      done();
    }).catch(done);
  });

  after(function(done) {
    this.server.stop(done);
  });

  it('server handles OPTION request type', function(done) {

  const options = {
    hostname: '127.0.0.1',
    port: 8080,
    path: '/happner-files/files/some/path/LICENSE',
    method: 'OPTIONS',
    headers: {
    }
  };

  const req = http.request(options, (res) => {
    res.should.have.property("statusCode",204)
    res.headers.should.have.property("access-control-allow-origin", "*")
    res.headers.should.have.property("access-control-allow-methods", "GET, PUT, OPTIONS")
    res.headers.should.have.property("access-control-allow-headers", "Content-Type, Content-Size, Access-Control-Allow-Origin")
    done()
  });

  req.on('error', (e) => {
    console.error(`problem with request: ${e.message}`);
    done(e)
  });
  req.end();
  });
});


test/func-upload-file-secure.js
// same as insecure, security in happn (preceding layer)

var path = require('path');

describe(path.basename(__filename), function() {
  it('');
});


example/static/script.js
function startPage() {

  var client;
  var username = document.getElementById('username');
  var password = document.getElementById('password');
  var login = document.getElementById('login');
  var logout = document.getElementById('logout');
  var loginStatus = document.getElementById('loginStatus');

  login.onclick = function() {
    loginStatus.innerHTML = 'pending login';
    client = new Happner.MeshClient();
    client.login({
      username: username.value,
      password: password.value
    });
    client.once('login/allow', function() {
      loginStatus.innerHTML = 'logged in as ' + username.value;
    });

    client.once('login/deny', function(error) {
      loginStatus.innerHTML = error.toString();
    });

    client.once('login/error', function(error) {
      loginStatus.innerHTML = error.toString();
    });
  };

  logout.onclick = function() {
    document.cookie = 'happn_token=; expires=Thu, 01 Jan 1970 00:00:01 GMT;';
    loginStatus.innerHTML = 'logged out';
  };


  var uploadPercentStatus = document.getElementById('uploadPercentStatus');
  var uploadStatus = document.getElementById('uploadStatus');
  var oem1File = document.getElementById('oem1-file');
  var oem2File = document.getElementById('oem2-file');

  var uploadFile = function(uploadToUrl, fileHandle) {
    uploadPercentStatus.innerHTML = 'uploaded 0%';
    uploadStatus.innerHTML = 'uploading';

    var request = new XMLHttpRequest();
    request.upload.onprogress = function(event) {
      var percent = (event.loaded / event.total * 100);
      uploadPercentStatus.innerHTML = 'uploaded ' + percent + '%';
    };
    request.onreadystatechange = function(e) {
      if(this.readyState === 4) {
        console.log('status', request.status);
        uploadStatus.innerHTML = 'upload done, statusCode ' + request.status;
      }
    };
    request.open('POST', uploadToUrl, true);
    request.send(fileHandle);


  };

  oem1File.onchange = function(event) {
    var fileHandle = event.target.files[0];
    var fileName = fileHandle.name;
    var uploadToUrl = '/happner-files/files/oem1/' + fileName;
    uploadFile(uploadToUrl, fileHandle);
  }

  oem2File.onchange = function(event) {
    var fileHandle = event.target.files[0];
    var fileName = fileHandle.name;
    var uploadToUrl = '/happner-files/files/oem2/' + fileName;
    uploadFile(uploadToUrl, fileHandle);
  }

}


lib/files.js
module.exports = Files;

var path = require('path');
var fs = require('fs');
var os = require('os');
var mkdirp = require('mkdirp');

function Files() {
  this.routes = [];
}

Files.prototype.handler = function($happn, req, res) {
  var targetFilename, targetFunction; // or directory (later)

  if (!this._setPathRoutes($happn)) {
    res.statusCode = 500;
    return res.end('happner-files: config missing path routes');
  }

  if (!(targetFilename = this._matchPathRoute(req.url))) {
    res.statusCode = 500;
    return res.end('happner-files: no matching path route');
  }

  targetFilename = this._transform(targetFilename);
  targetFunction = '_handle' + req.method;

  if (typeof this[targetFunction] !== 'function') {
    res.statusCode = 500;
    return res.end('happner-files: ' + req.method + ' method not implemented');
  }
  if(res.setHeader) {
    res.setHeader('Access-Control-Allow-Origin', '*');
    res.setHeader('Access-Control-Allow-Methods', 'GET, PUT, OPTIONS');
    res.setHeader(
      'Access-Control-Allow-Headers',
      'Content-Type, Content-Size, Access-Control-Allow-Origin'
    );
  }
  this[targetFunction]($happn, targetFilename, req, res);
};

Files.prototype._setPathRoutes = function($happn) {
  if (this.routes.length > 0) return true;
  if (!$happn.config.path) return false;
  if (!$happn.config.path.routes) return false;
  var routes = this.routes;
  Object.keys($happn.config.path.routes).forEach(function(key) {
    var regex = new RegExp('^' + key);
    var path = $happn.config.path.routes[key];
    routes.push({match: regex, path: path});
  });
  return true;
};

Files.prototype._matchPathRoute = function(url) {
  var targetFilename, i;
  for (i = 0; i < this.routes.length; i++) {
    if (!url.match(this.routes[i].match)) continue;
    targetFilename = url.replace(this.routes[i].match, this.routes[i].path + path.sep);
    if (os.platform() === 'win32') {
      targetFilename = targetFilename.replace(/\//g, '\\');
    }
    return path.normalize(targetFilename);
  }
  return false;
};

Files.prototype._transform = function(targetFilename) {
  return targetFilename;
};

Files.prototype._handlePOST = function($happn, targetFilename, req, res) {
  $happn.log.$$DEBUG('processing %s %s to %s', req.method, req.url, targetFilename);
  mkdirp(path.dirname(targetFilename), function(error) {
    if (error) {
      res.statusCode = 500;
      return res.end('happner-files: failed to create directory');
    }
    var stream = fs.createWriteStream(targetFilename);
    var errored = false;

    // stream.pipe(req); // why not?

    req.on('data', function(buf) {
      stream.write(buf);
    });

    req.on('end', function() {
      stream.end();
    });

    req.on('error', function(error) {
      $happn.log.warn('request error', error);
      res.statusCode = 500;
      res.end('happner-files: request error');
      stream.end();
      errored = true;
    });

    stream.on('error', function(error) {
      $happn.log.warn('stream error', error);
      res.statusCode = 500;
      res.end('happner-files: stream error');
      stream.end();
      errored = true;
    });

    stream.on('close', function() {
      if (errored) return;
      res.statusCode = 201;
      res.end();
    });
  });
};


Files.prototype._handleGET = function($happn, targetFilename, req, res) {
  $happn.log.$$DEBUG('processing %s %s to %s', req.method, req.url, targetFilename);
  fs.lstat(targetFilename, function(error, stat) {
    if (error) {
      if (error.code === 'ENOENT') {
        res.statusCode = 404;
        return res.end();
      }
      $happn.log.error('GET error', error);
      res.statusCode = 500;
      return res.end();
    }

    if (stat.isDirectory()) {
      res.statusCode = 404;
      return res.end();
    }

    var stream = fs.createReadStream(targetFilename);
    stream.pipe(res);
  });
};


Files.prototype._handleOPTIONS = function($happn, targetFilename, req, res) {
  res.statusCode = 204;
  return res.end();
};
